### DISCUSSION TITLE
RLS with joins

### DISCUSSION THREAD

**Discussion by adamhp** [supabase/supabase#4435] [Questions] [answered]
I've got two tables that are pretty straight forward. One is `locations` which stores location objects, and `location_images` which stores references to the locations, along with an image path that points to a storage object so I can pull from storage later.

I want to make it so that only users that "own" the particular locations can upload images for them. The `locations` table has an `owner_id` that says who owns it.

I tried creating the following policy:
```sql
CREATE POLICY "Enable insert for owned images" ON public.location_images 
  FOR INSERT WITH CHECK (
    auth.role() = 'authenticated' 
    AND 
    EXISTS ( 
      SELECT owner_id
      FROM public.locations
      WHERE owner_id = auth.uid()
    )
);
```

But I get RLS violations when I attempt to create entries in `location_images`. Am I formulating this query incorrectly?

**Comment by GaryAustin1** [comment, marked as answer]
Try putting
SELECT 1
FROM locations
WHERE locations.owner_id = auth.uid()

You don't show how your tables are set up with owner_id.

Also, If you are doing the insert with supabase.js (at least) you need to either turn set returning parameter to minimal or also have a RLS policy for select as the insert also does a select.  https://supabase.com/docs/reference/javascript/insert

An aside your check seems to be just saying if this user has an entry in the locations table he can put an entry in the locations_images table.  It does not on it's own enforce any sort of match to a specific location.

**Reply by adamhp** [reply]
Thanks for saving me again, Gary! I do think it was the SELECT policy. Haven't tested the ownership query just yet, but it should work. I'll reply back here once finished for a more complete explanation for others.

(12/16) Edit: I haven't forgotten about this, just haven't gotten around to this implementation!

(1/10) Edit: Still working on it! Apologies to anyone tuned into a similar problem.

(1/22) Edit: **SOLVED** Okay! Finally got around to it! The following policy works, but I've only tested happy path.

```postgresql
CREATE POLICY "Enable insert for owned images" ON public.location_images 
  FOR INSERT WITH CHECK (
    auth.role() = 'authenticated' 
    AND 
    EXISTS ( 
      SELECT 1 FROM locations 
      WHERE ((locations.owner_id = uid()) 
      AND (locations.id = location_images.location_id))
    )
  );
```

So this assumes that when you're inserting an object, it has to have a `location_id`, and that `location_id` has to exist in the locations table, and the owner of that particular location has to be the current user. Definitely verify this logic, as I've only tested happy path, but it seems to be working okay for me for now.
